main/src/library/MutHeap.flix

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2023 Xavier deSouza
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pub enum MutHeap[a: Type, r: Region] {
+    case MutHeap(Region[r], Ref[Array[a, r], r], Ref[Int32, r])
+}
+
+mod MutHeap {
+
+    ///
+    /// Returns an empty MutHeap
+    ///
+    pub def new(rc: Region[r]): MutHeap[a, r] \ r =
+        MutHeap(rc, ref Array.new(rc, 8) @ rc, ref 0 @ rc)
+
+    ///
+    /// Returns the number of elements in a MutHeap
+    ///
+    pub def size(h: MutHeap[a, r]): Int32 \ r =
+        let MutHeap(_, _, s) = h;
+        deref s
+
+    ///
+    /// Returns whether a MutHeap is empty (has 0 elements)
+    ///
+    pub def isEmpty(h: MutHeap[a, r]): Bool \ r =
+        let MutHeap(_, _, s) = h;
+        deref s == 0
+
+    ///
+    /// Returns the top element of a MutHeap
+    ///
+    pub def peek(h: MutHeap[a, r]): Option[a] \ r =
+        let MutHeap(_, arr, s) = h;
+        if (deref s == 0) None
+        else Some(Array.get(0, deref arr))
+
+    ///
+    /// Removes the top element of a MutHeap
+    ///
+    pub def pop(h: MutHeap[a, r]): Unit \ r with Order[a] =
+        let MutHeap(_, _, s) = h;
+        if (deref s > 0) {
+            match peek(h) {
+                case Some(elem) => delete(h, elem)
+                case _          => ()
+            }
+        }
+        else ()
+
+    ///
+    /// Inserts an element into a MutHeap
+    ///
+    pub def insert(h: MutHeap[a, r], elem: a): Unit \ r with Order[a] = {
+        let MutHeap(_, arr, s) = h;
+        if (deref s == 0) {
+            Array.put(elem, 0, deref arr);
+            s := 1
+        }
+        else {
+            expand!(h);
```

Comment:
Not specified because it's in a helper, but expend! optionally expands if the size is approaching the internal array's capacity. It won't do anything otherwise; this reflects some of the other mutable structures in the library